_posts/2018-6-11-Interpreter-Optimizations.md

```diff
@@ -0,0 +1,209 @@
+in the [internal beam\_makeops documentation]. The largest difference
+inbetween the new and old way of doing the code generation is that
+in OTP 21 _all_ the instructions are now generated, instead of about 75%.
+This allowed us to do architecture specific optimizations for all instructions
+when generating the code for the instructions.
+
+[interpreter]: https://en.wikipedia.org/wiki/Interpreter_(computing)
+[direct threading]: https://en.wikipedia.org/wiki/Threaded_code#Direct_threading
+[labels as values]: https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html
+[beam\_makeops]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/utils/beam_makeops
+[genops.tab]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/lib/compiler/src/genops.tab
+[ops.tab]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/beam/ops.tab
+[instrs.tab]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/beam/instrs.tab
+[internal beam\_makeops documentation]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/internal_doc/beam_makeops.md
+[transform\_engine]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/beam/beam_load.c#L5213
+
+## Packing the dispatch target address ##
+
+On 64-bit machines all pointers are 8 byte large, this is also the case for
+the pointer that you get when taking the address of a label. So for a small
+instruction such as move_cx (move a constant to a x register), 3 words of
+memory is needed.
+
+         +--------------------+--------------------+
+    I -> |                            &&lb_move_cx |
+         +--------------------+--------------------+
+         |                                      40 |
+         +--------------------+--------------------+
+         |                        Tagged atom 'id' |
+         +--------------------+--------------------+
+
+One word for the instruction, one word for the literal and then the target x
+register actually only needs 2 bytes, but since all code is word aligned it
+gets its own word.
+
+However, on most machines, the linker uses what is called a small or medium code model.
+This means that it will place all code in the lower 2 GB of the address space
+so that more efficient machine instructions can be used. It also works in the
+favour of the beam interpreter as we now know that the upper 4 bytes of the
+instruction word will always be 0.
+
+         +--------------------+--------------------+
+    I -> |                 40 |       &&lb_move_cx |
+         +--------------------+--------------------+
+         |                        Tagged atom 'id' |
+         +--------------------+--------------------+
+
+So instead of placing the x register to use in its own word we pack it into the
+instruction word, thus saving one word of memory for this instruction.
+
+The optimization is only possible to do on 64-bit platforms that place all code in the
+lower 4 GB of the address space, so that makes 32-bit and [pie] platforms not use it.
```

Comment:
For clarity, I suggest writing "position-independent executable platforms" in full.